src/dd/dd.rs

```diff
@@ -0,0 +1,745 @@
+    let mut total = 0;
+    for _ in 0..cf.count {
+        let mut pos = 0;
+        let mut left = cf.inp_block_size;
+        let mut session = 0;
+        'read_block: loop {
+            if pos == cf.inp_block_size {
+                break 'read_block;
+            }
+            let mut is_read_0 = false;
+            match inp.read(&mut buf[pos..pos + left]) {
+                Ok(0) => is_read_0 = true,
+                Ok(num) => {
+                    session += num;
+                    total += num;
+                    if num < cf.inp_block_size {
+                        cf.inp_part += 1;
+                        pos += num;
+                        left = cf.inp_block_size - pos;
+                    } else if num == cf.inp_block_size {
+                        cf.inp_full += 1;
+                    } else {
+                        err(format!("overread: {} over {}", num, cf.inp_block_size));
+                    }
+                }
+                Err(_) => {
+                    // TODO: handle
+                }
+            }
+            if session == cf.inp_block_size || is_read_0 {
+                break 'read_block;
+            }
+        }
+        let num = cmp::min(session, cf.inp_block_size);
+        let mut pos = 0;
+        loop {
+            if pos == num {
+                break;
+            }
+            if let Ok(m) = out.write(&buf[pos..pos + num]) {
+                if m < cf.out_block_size {
+                    cf.out_part += 1;
+                } else {
+                    cf.out_full += 1;
+                }
+                pos += m;
+            } else {
+                err("error occured".to_string()); // TODO: handle
+            }
+        }
+        if total >= cf.inp_block_size * cf.count {
+            break;
+        }
+    }
+    drop(cf);
+    display_stat(Arc::clone(&first));
+
+    Ok(())
+}
+
+enum InpFileType {
+    Reg(File),
+    Stdin(io::Stdin),
+}
+
+pub struct InpFileWrap {
+    f: InpFileType,
+    is_seekable: bool,
+}
+
+impl InpFileWrap {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match self.f {
+            InpFileType::Reg(ref mut f) => f.read(buf),
+            InpFileType::Stdin(ref mut s) => s.read(buf),
+        }
+    }
+}
+
+enum OutFileType {
+    Reg(File),
+    Stdout(io::Stdout),
+}
+
+pub struct OutFileWrap {
+    f: OutFileType,
+    is_seekable: bool,
+}
+
+impl OutFileWrap {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        match self.f {
+            OutFileType::Reg(ref mut f) => f.write(buf),
+            OutFileType::Stdout(ref mut s) => s.write(buf),
+        }
+    }
+}
+
+pub fn uumain(_args: Vec<String>) -> i32 {
+    let signal = chan_signal::notify(&[Signal::INT]);
+    let (s_done, r_done) = chan::sync(0);
+
+    let cf = parse_args();
+    let cf1 = Arc::new(Mutex::new(cf));
+    let cf2 = Arc::clone(&cf1);
+
+    let handle = thread::spawn(move || {
+        let cf3 = Arc::clone(&cf1);
```

Comment:
This clone isn't necessary.

Reply:
Sadly, I don't know how to do it another way. I clone because it's also used by signal handler,
so I wrap value in smart pointer and pass around copies of that Arc.

Reply:
You aren't using `cf1` anywhere else in the function though, so AFAICT you should be able to just use `cf1` directly.

Reply:
Fixed.